bootstrap_salt/fab_tasks.py

```diff
@@ -100,211 +105,121 @@ def application(application_name):
-                                               zone_name)
-    r53.update_dns_record(zone_id, dns_name, 'A', master_ip)
-    return True
 
+def get_instance_ips():
+    """
+    Get a list of the public IPs of the current instances in the stack.
+    """
+    cf = get_connection(Cloudformation)
+    ec2 = get_connection(EC2)
+    stack_name = get_stack_name()
+    instance_ids = cf.get_stack_instances(stack_name)
+    instance_ids = [i.id for i in instance_ids]
+    instance_ips = ec2.get_instance_public_ips(instance_ids)
+    return instance_ips
 
-def put_util_script():
-    # copy the salt_utils.py from local to EC2 and chmod it
-    d = os.path.dirname(__file__)
-    saltutils = d + "/salt_utils.py"
-    if not os.path.isfile(saltutils):
-        print "ERROR: Cannot find %s" % saltutils
-        sys.exit(1)
-    put(saltutils, '/usr/local/bin', use_sudo=True)
-    sudo('chmod 755 /usr/local/bin/salt_utils.py')
 
+@task
+def wait_for_minions(timeout=600, interval=20):
+    """
+    This task ensures that the initial bootstrap has finished on all
+    stack instances.
 
-def install_minions(salt_version):
+    Args:
+        timeout(int): time to wait for bootstrap to finish
+        interval(int): time to wait inbetween checks
+    """
     _validate_fabric_env()
     stack_name = get_stack_name()
     ec2 = get_connection(EC2)
     print "Waiting for SSH on all instances..."
     ec2.wait_for_ssh(stack_name)
+    fab_hosts = get_instance_ips()
+    utils.timeout(timeout, interval)(is_bootstrap_done)(fab_hosts)
 
-    master_inst = ec2.get_master_instance(stack_name)
-    master_public_ip = master_inst.ip_address
-    master_prv_ip = master_inst.private_ip_address
-
-    to_install = ec2.get_unconfigured_minions(stack_name, master_prv_ip)
-
-    if not to_install:
-        return
-
-    sha = '6080a18e6c7c2d49335978fa69fa63645b45bc2a'
-    for instance in to_install:
 
-        env.host_string = 'ubuntu@%s' % instance.ip_address
+def is_bootstrap_done(hosts):
+    """
+    Loops through a list of IPs and checks for the prescence of
+    /tmp/bootstrap_done to ensure that the launch config has finished
+    executing.
 
-        put_util_script()
-        run('wget https://raw.githubusercontent.com/saltstack/salt-bootstrap/%s/bootstrap-salt.sh -O /tmp/bootstrap-salt.sh' % sha)
-        sudo('chmod 755 /tmp/bootstrap-salt.sh')
-        sudo('/tmp/bootstrap-salt.sh -A {0} -p python-boto git '
-             '{1}'.format(master_prv_ip, salt_version))
-        env.host_string = 'ubuntu@%s' % master_public_ip
-        sudo('salt-key -y -A')
+    Args:
+        hosts(list): A list of IPs to check
+    """
+    with settings(warn_only=True):
+        ret = []
+        for host in hosts:
+            env.host_string = '{0}@{1}'.format(env.user, host)
+            res = run('[ -f /tmp/bootstrap_done ]')
+            if res.return_code == 0:
+                ret.append(True)
+            else:
+                ret.append(False)
```

Comment:
Could we use http://docs.fabfile.org/en/1.10/api/contrib/files.html?highlight=exists#fabric.contrib.files.exists here?